H1 `CONNECT-UDP` support

deweerdt opened the pull request:
This PR adds support for `CONNECT-UDP` to the existing connect handler.
It also adds support for udp tunnels to the HTTP/1 stack.

kazuho commented on include/h2o/httpclient.h:
```diff
@@ -173,6 +173,30 @@ typedef struct st_h2o_httpclient_timings_t {
+    /**
+     * The write callback. Contrary to what happens in the tcp tunnel, completion is immediate.
+     */
+    void (*write_)(struct st_h2o_httpclient_udp_tunnel_t *tunnel, const void *bytes, size_t len);
```
I think we'd want to pass list of vectors here (like `writev`). UDP packets tend to be small, and there's performance benefit by exchanging them with the kernel in batch.

kazuho commented on include/h2o/httpclient.h:
```diff
@@ -173,6 +173,30 @@ typedef struct st_h2o_httpclient_timings_t {
+    /**
+     * the on-read callback to be set by the user
+     */
+    void (*on_read)(struct st_h2o_httpclient_udp_tunnel_t *tunnel, const char *err, const void *bytes, size_t len);
```
Same as above.

deweerdt replied:
Thanks for the suggestion. For reads, are you suggesting we use `recvmmsg` or is your idea that we change that signature to use iovecs?

kazuho replied:
It's more about the tunnel API, to make sure that we have room for optimization (to send / receive packets in batch). Such optimizations could be sendmmsg / recvmmsg, UDP GSO / GRO, io_uring, or whatever.

As not all platforms have sendmmsg / recvmmsg, I do not think it's necessary for us to use the -**m**msg variant at this moment. Though doing a loop for at most 10 packets or something would not be a bad idea.

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,106 @@
+        if (bind(fd, (void *)&sin, sizeof(sin)) != 0) {
+            perror("failed to bind bind UDP socket");
+            exit(EXIT_FAILURE);
+        }
```
I'm not sure if we want to use an unconnected socket.

Unconnected socket has three notable properties:
* can receive from any address
* can send to any address
* send path is slower

For CONNECT-UDP, we do not need the first two properties, and therefore, it would make sense to use a connect socket to avoid the third property.

Separately, I think we can move the creation of the socket to the open function? I'm not sure if there's benefit in delaying the operation.

kazuho commented on lib/handler/connect.c:
```diff
@@ -152,9 +168,15 @@ static int on_req(h2o_handler_t *_handler, h2o_req_t *req)
+    if (h2o_memis(req->input.method.base, req->input.method.len, H2O_STRLIT("CONNECT")))
+        connect_type = CONNECT;
+    else if (h2o_memis(req->input.method.base, req->input.method.len, H2O_STRLIT("CONNECT-UDP")))
+        connect_type = CONNECT_UDP;
```
maybe do `else return -1` here? Then, we can get rid of NONE.

kazuho commented on lib/handler/connect.c:
```diff
@@ -174,8 +197,8 @@ static int on_req(h2o_handler_t *_handler, h2o_req_t *req)
-                                             h2o_iovec_init(creq->server_name.port, port_strlen), AF_UNSPEC, SOCK_STREAM,
-                                             IPPROTO_TCP, AI_ADDRCONFIG | AI_NUMERICSERV, on_getaddr, creq);
+            h2o_iovec_init(creq->server_name.port, port_strlen), AF_UNSPEC, SOCK_STREAM,
+            IPPROTO_TCP, AI_ADDRCONFIG | AI_NUMERICSERV, on_getaddr, creq);
```
I believe that SOCK_STREAM and IPPROTO_TCP should be something else for UDP. And, assuming that these values can be obtained from one of the arguments passed to `on_getaddr` callback, we can get rid of `connect_type`.

kazuho commented on lib/http1.c:
```diff
@@ -92,8 +92,17 @@ struct st_h2o_http1_chunked_entity_reader {
+    union {
+        h2o_httpclient_tunnel_t *tcp;
+        h2o_httpclient_udp_tunnel_t *udp;
+    } server;
```
These changes makes me wonder if it is a good idea to introduce `h2o_httpclient_udp_tunnel_t` as something separate from `h2o_httpclient_tunnel_t`.

From the perspectives of HTTP/1 and HTTP/2 implementation, only "streams" can carry UDP datagrams. Even for HTTP/3, use of stream is the default; use of QUIC datagram frames is an "option." Regardless of the HTTP version being used, encapsulation format of UDP datagrams sent over the HTTP streams of respective versions is going to be the same.

Considering these aspects, would it make sense to merge the definion of `h2o_httpclient_udp_tunnel_t` into `h2o_httpclient_tunnel_t`, and letting the tunnel handle encapsulation?

I mean, h2o_httpclient_tunnel_t can have an optional `on_udp_read` callback. When it is set by the H3 server-side implementation, the tunnel would provide the UDP datagrams as they are through the callback. OTOH, when the callback is set to NULL, h2o_httpclient_tunnel_t would provide the encapsulated byte stream through the `on_read` callback.

For the write side, h2o_httpclient_tunnel_t can always provide both interfaces (i.e. `write_` and `udp_write`) assuming that a UDP tunnel is being opened. It would be the responsibility of the tunnel to unpack the encapsulated payload that it receives through the `write_` callback.

deweerdt replied:
that makes sense to me, thank you. What do you think of moving all the code (tcp or udp) to lib/common/tunnel.c?

deweerdt replied:
Mmm, i see now that it would make sense to keep both tunnels in their current files, there might be a handful of shared functions, but it'll be easier to split the implementations in different modules.

kazuho replied:
I do not have a concrete opinion, but I am not sure how much code can be shared, based on my assumption that the buffering strategy would be different. In case of TCP, the tunnel cannot drop anything, while in case of UDP, the tunnel is expected to have a fixed size maximum.

I would be fine with splitting the two, especially when the amount of logic that can be shared would be less than the amount of the logic that could be shared on the user-side; our H1 and H3 server-side implementations both use the tunnel API, they do have some code that look similar, but I tend to think that not trying to share stuff is giving us more benefit than drawbacks.

kazuho reviewed (commented):
Thank you for the PR. Looks promising, I have some thoughts around the design. Maybe the most significant one is for lib/http1.c. Please let me know what you think.

deweerdt commented:
@kazuho i believe that ecf5e85 addresses all your comments so far. It made `on_read` in the tunnel take iovecs, it also turns out that the whole tcp vs udp distinction wasn't necessary once the deserialization is done within the tunnel.

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+    return;
+}
+
+h2o_iovec_t get_next_chunk(const uint8_t *bytes, size_t len, size_t *to_consume, int *skip)
```
I think we can reduced the information being returned to the caller from three elements to two elements, by changing the signature of the function to: `size_t get_next_chunk(const uint8_t *bytes, size_t len, h2o_iovec_t *chunk)`.

The only case where the vector has to be examined is when some bytes are to be consumed. And a NULL chunk with some bytes being consumed can indicate the fact that a chunk has been skipped.

kazuho commented on include/h2o/tunnel.h:
```diff
@@ -52,14 +52,19 @@ typedef struct st_h2o_tunnel_t {
      * The on-read callback to be set by the user.
      */
-    void (*on_read)(struct st_h2o_tunnel_t *tunnel, const char *err, const void *bytes, size_t len);
+    void (*on_read)(struct st_h2o_tunnel_t *tunnel, const char *err, h2o_iovec_t *iov, size_t iovlen);
```
I am not sure if we need this change. Please see below.

kazuho replied:
Addressed in cc327b1 (#2568).

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+        vec[1].base = (char *)varint.buf;
+        vec[1].len = varint.end - varint.buf;
+        vec[2] = h2o_iovec_init(iov.iov_base, iov.iov_len);
+        tunnel->super.on_read(&tunnel->super, NULL, vec, sizeof(vec) / sizeof(vec[0]));
```
We cannot pass a buffer that points to local variables. That is because the bytes being supplied must be retained by the tunnel until the user of the tunnel calls `proceed_read`. Therefore, the UDP payload and the UDP chunk headers have to be within the heap instead of the stack.

One intuitive approach would be to have a fixed size buffer in `st_h2o_udp_tunnel_t`. `recvmsg` would read bytes directly into that buffer. When `on_udp_read` is not available, we would reverse three bytes of space before each call to `recvmsg`, write the UDP chunk header length after `recvmsg` returns (we can be sure that the size of each UDP datagram is below 16384 bytes, and we can spend one extra byte when encoding UDP datagram size below 64 bytes). If we use an approach like this, there's no need to pass a vector of vectors to `on_read`.

kazuho replied:
Addressed in cc327b1 (#2568).

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+        vec[i] = (struct iovec){
+            .iov_base = iov[i].base,
+            .iov_len = iov[i].len,
+        };
```
I'm not sure if this is correct. IIUC, each `iov` is a UDP datagram. Assuming that is the case, I think `sendmsg` should be called for each `iov` element?

PS. And assuming that is the case, I think we should refrain from calling the arguments "iov."

kazuho replied:
Addressed in 40e3399 (#2568).

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+        if (iovs[iovlen].len == 0)
+            break;
+        if (!skip)
+            iovlen++;
```
I think we might be concerned of overruns?

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+    struct st_h2o_udp_tunnel_t *tunnel = (void *)_tunnel;
+    h2o_iovec_t iovs[64];
+    size_t iovlen = 0;
+    size_t idx = 0;
```
How about calling this `off` or `offset`, as it is an index of a byte array?

kazuho replied:
Addressed in 40e3399 (#2568).

kazuho reviewed (commented):
Thank you for the changes. The design looks fine overall, I have couple of comments. PTAL.

kazuho commented on lib/common/udp_tunnel.c:
```diff
@@ -0,0 +1,208 @@
+                .proceed_read = tunnel_proceed_read,
+                .on_read = NULL,
+            },
+        .len = len,
```
Addressed in 8b83a2c (#2568).

kazuho commented on lib/http3/server.c:
```diff
@@ -1273,24 +1273,27 @@ static void do_send_informational(h2o_ostream_t *_ostr, h2o_req_t *_req)
+        stream->sendbuf.final_body_size += iov[i].len;
         /* update final offset */
-        stream->sendbuf.final_size += header_size + len;
+        stream->sendbuf.final_size += header_size + iov[i].len;
```
Addressed in commits up to 7033bfa (#2568).

kazuho commented:
This PR has been subsumed by #2568.
